src/org/jitsi/impl/neomedia/rtp/translator/RTCPFeedbackMessageSender.java

```diff
@@ -102,87 +113,292 @@ private long getSenderSSRC()
+         */
+        private final int mediaSenderSSRC;
+
+        /**
+         * The sequence number of the next FIR.
+         */
+        private final AtomicInteger sequenceNumber = new AtomicInteger(0);
+
+        /**
+         * The number of FIR that are left to be sent before stopping.
+         */
+        private int remainingRetries;
+
+        /**
+         * Ctor.
+         *
+         * @param mediaSenderSSRC
+         */
+        public KeyframeRequester(int mediaSenderSSRC)
         {
-            if (sendFIR(destination, mediaSenderSSRC))
-                sentFIR = true;
+            super(FIR_RETRY_INTERVAL_MS);
+            this.mediaSenderSSRC = mediaSenderSSRC;
+            this.remainingRetries = 0;
+        }
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public long process()
+        {
+            super.process();
+
+            this.maybeRequest(false);
+
+            return 0; /* unused */
+        }
+
+        /**
+         * Notifies this instance that an RTP packet has been received from a
+         * peer represented by a specific <tt>StreamRTPManagerDesc</tt>.
+         *
+         * @param streamRTPManager a <tt>StreamRTPManagerDesc</tt> which
+         * identifies the peer from which an RTP packet has been received
+         * @param buf the buffer which contains the bytes of the received RTP or
+         * RTCP packet
+         * @param off the zero-based index in <tt>buf</tt> at which the bytes of
+         * the received RTP or RTCP packet begin
+         * @param len the number of bytes in <tt>buf</tt> beginning at
+         * <tt>off</tt> which represent the received RTP or RTCP packet
+         */
+        public void maybeStopRequesting(
+            StreamRTPManagerDesc streamRTPManager,
+            int pt,
+            byte[] buf,
+            int off,
+            int len)
+        {
+            if (remainingRetries == 0)
+            {
+                return;
+            }
+
+            // Reduce auto-boxing (even tho the compiler or the JIT should do
+            // this automatically).
+            Byte redPT = null, vp8PT = null;
+
+            // XXX do we want to do this only once?
+            for (Map.Entry<Byte, MediaFormat> entry : streamRTPManager
+                .streamRTPManager.getMediaStream().getDynamicRTPPayloadTypes()
+                .entrySet())
+            {
+                String encoding = entry.getValue().getEncoding();
+                if (Constants.VP8.equals(encoding))
+                {
+                    vp8PT = entry.getKey();
+                }
+                else if (Constants.RED.equals(encoding))
+                {
+                    redPT = entry.getKey();
+                }
+            }
+
+            if (vp8PT == null || vp8PT != pt)
+            {
+                return;
+            }
+
+            RawPacket pkt = new RawPacket(buf, off, len);
+            if (!Utils.isKeyFrame(pkt, redPT, vp8PT))
```

Comment:
You're right, it doesn't make any sense to allocate a `RawPacket` here. Thank you @lyubomir, I'll fix it!